### ISSUE TITLE
/bin/sh: can't access tty; job control turned off

### ISSUE THREAD

**Issue by joerg-krause** [troglobit/finit#91]
I am testing 3.1-rc1 and I am getting the message `/bin/sh: can't access tty; job control turned off` when booting with the following config file on a Banana Pro using Buildroot with Busybox (replaced `init` by `finit`:

```
tty [12345] /sbin/getty -L 115200 /dev/ttyS0 vt100
```

Also tried:

```
tty [12345] @console 115200 linux noclear
```

Note, that finit is not yet supported by Buildroot, but I am working on it :smile: For now, I am quite unsure if this is an issue or not.

**Comment by troglobit** [comment]
I'm guessing you get this message after having logged in.  It usually happens only when there's something wrong with `/dev`, because the built-in getty (your second try above) calls `ioctl(STDIN_FILENO, TIOCSCTTY, 1);` to enable job control in the TTY.  I'm guessing your `/sbin/getty` tries something similar and if `/dev/ is not in shape that might happen.

Finit expects the kernel to handle `/dev/`, so the kernel needs to be built with:

```
CONFIG_DEVTMPFS=y
CONFIG_DEVTMPFS_MOUNT=y
```

This is a limitation that sometimes comes up as a hindrance to adoption since not all vendor kernels have this enabled. Not sure, however, if this is what you've run in to though ... 

Anyway, glad to hear someone trying out Finit with Buildroot :-)

**Comment by joerg-krause** [comment]
I've changed the configuration and built a Buildroot system without Busybox (and getty), with devtmpfs and dash (`/bin/sh` is a symlink to dash). The systems does not have `/bin/login`.

The Linux kernel is configured with:
```
CONFIG_DEVTMPFS=y
CONFIG_DEVTMPFS_MOUNT=y
```

`/dev` looks okay to me:

```
# cd /dev
# echo *
bus console cpu_dma_latency fb0 full gpiochip0 i2c-0 i2c-1 iio:device0 input kmsg mem memory_bandwidth mmcblk0 mmcblk0p1 mmcblk0p2 network_latency network_throughput null ptmx ptya0 ptya1 ptya2 ptya3 ptya4 ptya5 ptya6 ptya7 ptya8 ptya9 ptyaa ptyab ptyac ptyad ptyae ptyaf ptyb0 ptyb1 ptyb2 ptyb3 ptyb4 ptyb5 ptyb6 ptyb7 ptyb8 ptyb9 ptyba ptybb ptybc ptybd ptybe ptybf ptyc0 ptyc1 ptyc2 ptyc3 ptyc4 ptyc5 ptyc6 ptyc7 ptyc8 ptyc9 ptyca ptycb ptycc ptycd ptyce ptycf ptyd0 ptyd1 ptyd2 ptyd3 ptyd4 ptyd5 ptyd6 ptyd7 ptyd8 ptyd9 ptyda ptydb ptydc ptydd ptyde ptydf ptye0 ptye1 ptye2 ptye3 ptye4 ptye5 ptye6 ptye7 ptye8 ptye9 ptyea ptyeb ptyec ptyed ptyee ptyef ptyp0 ptyp1 ptyp2 ptyp3 ptyp4 ptyp5 ptyp6 ptyp7 ptyp8 ptyp9 ptypa ptypb ptypc ptypd ptype ptypf ptyq0 ptyq1 ptyq2 ptyq3 ptyq4 ptyq5 ptyq6 ptyq7 ptyq8 ptyq9 ptyqa ptyqb ptyqc ptyqd ptyqe ptyqf ptyr0 ptyr1 ptyr2 ptyr3 ptyr4 ptyr5 ptyr6 ptyr7 ptyr8 ptyr9 ptyra ptyrb ptyrc ptyrd ptyre ptyrf ptys0 ptys1 ptys2 ptys3 ptys4 ptys5 ptys6 ptys7 ptys8 ptys9 ptysa ptysb ptysc ptysd ptyse ptysf ptyt0 ptyt1 ptyt2 ptyt3 ptyt4 ptyt5 ptyt6 ptyt7 ptyt8 ptyt9 ptyta ptytb ptytc ptytd ptyte ptytf ptyu0 ptyu1 ptyu2 ptyu3 ptyu4 ptyu5 ptyu6 ptyu7 ptyu8 ptyu9 ptyua ptyub ptyuc ptyud ptyue ptyuf ptyv0 ptyv1 ptyv2 ptyv3 ptyv4 ptyv5 ptyv6 ptyv7 ptyv8 ptyv9 ptyva ptyvb ptyvc ptyvd ptyve ptyvf ptyw0 ptyw1 ptyw2 ptyw3 ptyw4 ptyw5 ptyw6 ptyw7 ptyw8 ptyw9 ptywa ptywb ptywc ptywd ptywe ptywf ptyx0 ptyx1 ptyx2 ptyx3 ptyx4 ptyx5 ptyx6 ptyx7 ptyx8 ptyx9 ptyxa ptyxb ptyxc ptyxd ptyxe ptyxf ptyy0 ptyy1 ptyy2 ptyy3 ptyy4 ptyy5 ptyy6 ptyy7 ptyy8 ptyy9 ptyya ptyyb ptyyc ptyyd ptyye ptyyf ptyz0 ptyz1 ptyz2 ptyz3 ptyz4 ptyz5 ptyz6 ptyz7 ptyz8 ptyz9 ptyza ptyzb ptyzc ptyzd ptyze ptyzf random rtc0 snd tty tty0 tty1 tty10 tty11 tty12 tty13 tty14 tty15 tty16 tty17 tty18 tty19 tty2 tty20 tty21 tty22 tty23 tty24 tty25 tty26 tty27 tty28 tty29 tty3 tty30 tty31 tty32 tty33 tty34 tty35 tty36 tty37 tty38 tty39 tty4 tty40 tty41 tty42 tty43 tty44 tty45 tty46 tty47 tty48 tty49 tty5 tty50 tty51 tty52 tty53 tty54 tty55 tty56 tty57 tty58 tty59 tty6 tty60 tty61 tty62 tty63 tty7 tty8 tty9 ttyS0 ttyS1 ttyS2 ttyS3 ttyS4 ttyS5 ttyS6 ttyS7 ttya0 ttya1 ttya2 ttya3 ttya4 ttya5 ttya6 ttya7 ttya8 ttya9 ttyaa ttyab ttyac ttyad ttyae ttyaf ttyb0 ttyb1 ttyb2 ttyb3 ttyb4 ttyb5 ttyb6 ttyb7 ttyb8 ttyb9 ttyba ttybb ttybc ttybd ttybe ttybf ttyc0 ttyc1 ttyc2 ttyc3 ttyc4 ttyc5 ttyc6 ttyc7 ttyc8 ttyc9 ttyca ttycb ttycc ttycd ttyce ttycf ttyd0 ttyd1 ttyd2 ttyd3 ttyd4 ttyd5 ttyd6 ttyd7 ttyd8 ttyd9 ttyda ttydb ttydc ttydd ttyde ttydf ttye0 ttye1 ttye2 ttye3 ttye4 ttye5 ttye6 ttye7 ttye8 ttye9 ttyea ttyeb ttyec ttyed ttyee ttyef ttyp0 ttyp1 ttyp2 ttyp3 ttyp4 ttyp5 ttyp6 ttyp7 ttyp8 ttyp9 ttypa ttypb ttypc ttypd ttype ttypf ttyq0 ttyq1 ttyq2 ttyq3 ttyq4 ttyq5 ttyq6 ttyq7 ttyq8 ttyq9 ttyqa ttyqb ttyqc ttyqd ttyqe ttyqf ttyr0 ttyr1 ttyr2 ttyr3 ttyr4 ttyr5 ttyr6 ttyr7 ttyr8 ttyr9 ttyra ttyrb ttyrc ttyrd ttyre ttyrf ttys0 ttys1 ttys2 ttys3 ttys4 ttys5 ttys6 ttys7 ttys8 ttys9 ttysa ttysb ttysc ttysd ttyse ttysf ttyt0 ttyt1 ttyt2 ttyt3 ttyt4 ttyt5 ttyt6 ttyt7 ttyt8 ttyt9 ttyta ttytb ttytc ttytd ttyte ttytf ttyu0 ttyu1 ttyu2 ttyu3 ttyu4 ttyu5 ttyu6 ttyu7 ttyu8 ttyu9 ttyua ttyub ttyuc ttyud ttyue ttyuf ttyv0 ttyv1 ttyv2 ttyv3 ttyv4 ttyv5 ttyv6 ttyv7 ttyv8 ttyv9 ttyva ttyvb ttyvc ttyvd ttyve ttyvf ttyw0 ttyw1 ttyw2 ttyw3 ttyw4 ttyw5 ttyw6 ttyw7 ttyw8 ttyw9 ttywa ttywb ttywc ttywd ttywe ttywf ttyx0 ttyx1 ttyx2 ttyx3 ttyx4 ttyx5 ttyx6 ttyx7 ttyx8 ttyx9 ttyxa ttyxb ttyxc ttyxd ttyxe ttyxf ttyy0 ttyy1 ttyy2 ttyy3 ttyy4 ttyy5 ttyy6 ttyy7 ttyy8 ttyy9 ttyya ttyyb ttyyc ttyyd ttyye ttyyf ttyz0 ttyz1 ttyz2 ttyz3 ttyz4 ttyz5 ttyz6 ttyz7 ttyz8 ttyz9 ttyza ttyzb ttyzc ttyzd ttyze ttyzf urandom vcs vcs1 vcsa vcsa1 watchdog watchdog0 zero
```

To be honest, I do not know what job control in the TTY is about...

**Comment by troglobit** [comment]
I see, well you might want to try adding `/bin/login`, otherwise you likely may have to use the BusyBox `/bin/cttyhack` applet.  I've never tried this myself but something like the following may work:

    service [12345] /bin/cttyhack /bin/sh noclear nowait

You can try it out with 

    # exec cttyhack sh

from your current prompt.  You can also try removing all `tty` configs and build Finit with `--enable-fallback-shell`, but I'd highly recommend using `/bin/login` instead.  With a default "none" password for the root user its fairly usable for development purposes.

**Comment by troglobit** [comment]
OK, admittedly I've been in your shoes from time to time, and used various hacks to get around the issue.  However, in commit 72bc3b1 on the master branch I've added the following functionality to be able to bypass `/bin/login` and not even use a getty, neither the built-in nor an external one:

Put the following in your `/etc/finit.conf`, `/etc/finit.d/getty.conf`, or similar:

    tty [12345] @console noclear nologin

Thenew `nologin` flag bypasses getty + login and simply starts `/bin/sh` when you press Enter.  (You can bypass the latter as well with the `nowait` flag)

Merry Christmas! :santa:

**Comment by joerg-krause** [comment]
Thanks! The `nologin` feature is quite handy for development... However, the `can't access tty` message remains when using dash, but not when using bash. Note, that I am not concerned about the message at all, I do not even know what it means. I am just playing with the different Buildroot system configurations to see how Finit can be integrated into Buildroot. 

Merry Christmas, too!

**Comment by troglobit** [comment]
That is an odd message, with `nologin` Finit should take care of that by taking control of the TTY and enabling signals.  Which is the most annoying thing with that faulty state.  E.g., you try `ping  8.8.8.8` and realize Ctrl-C doesn't work ... 

... I'm reopening this issue (GitHub automatically closes issues if one mentions Fix #NUM in the commit message), so we can keep discussing this. I've tried out a sort-of bare-bones setup like yours in my very limited [TroglOS](https://github.com/troglobit/troglos), and there it works. Unfortunately its master branch is still not usable right now, but maybe you can find some pointers: e.g. kernel .config, busybox .config, etc.

Now, back to Christmas celebrations :)

**Comment by troglobit** [comment]
Discussions died out in this thread. Meanwhile, at $DAYJOB we've now adopted Buildroot for our [NetBox](https://github.com/westermo/netbox) platform (a buildroot external).  I've integrated the latest Finit and it works well, without any of the above warning messages.  A new version is also coming soon.  I suggest taking a look at how we've integrated Finit, if you're at all still interested.

Cheers
 /Joachim
